crates/sui-graphql-rpc/src/types/epoch.rs

```diff
@@ -259,13 +271,31 @@ impl Epoch {
     /// Look up an `Epoch` in the database, optionally filtered by its Epoch ID. If no ID is
     /// supplied, defaults to fetching the latest epoch.
     pub(crate) async fn query(
-        db: &Db,
+        ctx: &Context<'_>,
         filter: Option<u64>,
         checkpoint_viewed_at: Option<u64>,
+    ) -> Result<Option<Self>, Error> {
+        if let Some(epoch_id) = filter {
+            let dl: &DataLoader<Db> = ctx.data_unchecked();
+            dl.load_one(EpochKey {
+                epoch_id,
+                checkpoint_viewed_at,
+            })
+            .await
+        } else {
```

Comment:
It seems that here we choose not to use the data loader pattern when epoch id is not provided. I understand that in the `if` case, data loader helps because the loader can help us combine multiple queries to the db into one, thus improving the performance. Is it not applicable to the `else` case? In general, what's our strategy for determining if something is worthy of a `Loader` impl? I imagine the degree of potential performance improvement being one big factor. Are there other factors?

Reply:
> In general, what's our strategy for determining if something is worthy of a `Loader` impl?

Bear in mind that this is just the first time we've implemented a loader so maybe our understanding of these rules will change, but here's the set of rules we talked about when considering this trade-off:

- How often is this query made, particularly in sub-queries.
- How easy and efficient to implement is the batch version of the query.

We should consider loader implementations for queries that are easy to batch and happen often under sub-queries. In general this can leave some grey areas but in this case it's pretty clear cut:

- The only place where we offer a query on the "latest" epoch (modulo a checkpoint) is at the top-level, so it is unlikely to be called repeatedly, and never under a sub-query.
- Writing a multi-get query by ID is easy, whereas latest at checkpoint is not: The query you would have to write for "latest at checkpoint" is not going to be more efficient than `n` queries for `n` different checkpoints, really.

In general, I think this rule boils down to us having data loaders for any point lookups (look-up an entity by its ID) and that's it.